### PR TITLE
Add command to check ocis backup consistency

### PR DESCRIPTION
**Pull request by kobergj**
Adds consistency check command

Fixes https://github.com/owncloud/ocis/issues/9004
E2E testing needs additional ticket: https://github.com/owncloud/ocis/issues/9286

### CODE REVIEW

**Comment by kobergj** [comment]
Example output inconsistent case:
```
🚨 Inconsistent Nodes:
        👉️ [symlink missing]    path: /home/jkoberg/.ocis/storage/users/spaces/23/ebf113-76d4-43c0-8594-df974b02cd74/nodes/91/13/a7/18/-8285-4b32-9042-f930f1a58ac2
        👉️ [symlink missing]    path: /home/jkoberg/.ocis/storage/users/spaces/23/ebf113-76d4-43c0-8594-df974b02cd74/nodes/99/bd/9c/46/-0c9d-4e94-939b-0f0ab1d39238
🚨 Inconsistent Links:
        👉️ [node missing]       path: /home/jkoberg/.ocis/storage/users/spaces/6b/15688a-f0d8-47a2-b345-470181c4f016/nodes/d7/78/fc/16/-d3c0-490c-8866-9ab05e153a94
🚨 Inconsistent Blobs:
        👉️ [blob orphaned]      blob: d4174a97-3d9b-4529-aa25-cf5fd2029b1b
        👉️ [blob orphaned]      blob: 5f463b1e-5820-4240-bdea-2f3174f573a1
🚨 Inconsistent BlobReferences:
        👉️ [blob missing]       blob: 87b2a98f-9b27-4c9e-9bcd-7bcc55fed0aa
```

**Comment by mmattel** [comment]
Some documentation is needed. Dont go the route without...
In owncloud.dev, create a subsection like `Backup` and add all backup related content underneath, starting with the existing Backip Considerations documentation. See the image how it currently looks:

image

File: docs/ocis/backup.md [hunk 1]
```diff
@@ -217,3 +217,12 @@ BACKUP RECOMMENDED/OMITABLE. This folder contains custom web assets. Can be spec
...
 
+## Backup Consistency command
+
+Infinite Scale now allows checking an existing backup for consistency. Use command
+```bash
+ocis backup consistency -p "<path-to-base-folder>"
+```
+
+`path-to-base-folder` needs to be replaced with the path to the storage providers base path. For example: "/var/lib/ocis/storage/users" or "$HOME/storage/users".
+Use the `-b s3ng` when using an external (s3) blobstore. Note: When using this flag the blobstore must be configured via envvars or yaml file the same as in ocis. Consistency checks for other blobstores than `ocis` and `s3ng` are not supported at the moment.
```
Comment (mmattel) [inline docs/ocis/backup.md:228]:
```suggestion

Use the `-b s3ng` option when using an external (s3) blobstore. Note: When using this flag, the path to the blobstore must be configured via envvars or a yaml file to match the configuration of the original instance. Consistency checks for other blobstores than `ocis` and `s3ng` are not supported at the moment.
```

File: docs/ocis/backup.md [hunk 2]
```diff
@@ -217,3 +217,12 @@ BACKUP RECOMMENDED/OMITABLE. This folder contains custom web assets. Can be spec
 
 When using an external idp/idm/nats or blobstore, its data needs to be backed up separately. Refer to your idp/idm/nats/blobstore documentation for backup details.
 
+## Backup Consistency command
```
Comment (mmattel) [inline docs/ocis/backup.md:220]:
```suggestion
## Backup Consistency Command
```

File: docs/ocis/backup.md [hunk 3]
```diff
@@ -217,3 +217,12 @@ BACKUP RECOMMENDED/OMITABLE. This folder contains custom web assets. Can be spec
...
+
+Infinite Scale now allows checking an existing backup for consistency. Use command
```
Comment (mmattel) [inline docs/ocis/backup.md:222]:
```suggestion
Infinite Scale now allows checking an existing backup for consistency. Use the command:
```

File: docs/ocis/backup.md [hunk 4]
```diff
@@ -217,3 +217,12 @@ BACKUP RECOMMENDED/OMITABLE. This folder contains custom web assets. Can be spec
...
+```bash
+ocis backup consistency -p "<path-to-base-folder>"
+```
+
+`path-to-base-folder` needs to be replaced with the path to the storage providers base path. For example: "/var/lib/ocis/storage/users" or "$HOME/storage/users".
```
Comment (mmattel) [inline docs/ocis/backup.md:227]: `...the path to the storage providers base path.` This sentence makes the path origin not clear, you should add the relevant envvar for reference so one can have a look how it was defined.

Comment (kobergj) [reply docs/ocis/backup.md:227]: There are multiple envvars for it. Could be `BASE_DATA_BASE` or `STORAGE_USERS_OCIS_ROOT` or even `STORAGE_SYSTEM_OCIS_ROOT`. That is why put two examples instead.

Comment (mmattel) [reply docs/ocis/backup.md:227]: The docs say

STORAGE_USERS_OCIS_ROOT

The directory where the filesystem storage will store blobs and metadata. If not defined, the root directory derives from $OCIS_BASE_DATA_PATH:/storage/users.

Referencing an envvar is imho much clearer as we have very good docs about...

Comment (kobergj) [inline docs/ocis/backup.md:227]:
```suggestion
`path-to-base-folder` needs to be replaced with the path to the storage providers base path. Should be same as the `STORAGE_USERS_OCIS_ROOT`
```

**Comment by wkloucek** [comment]
@kobergj I gave this tool a try today. I was executing it in a environment that is configured for the storage-users service only (storageusers pod in the oCIS Helm Chart). To be actually able to run this tool, I needed to configure following environment variables:
- OCIS_JWT_SECRET
- OCIS_MACHINE_AUTH_API_KEY
- OCIS_SYSTEM_USER_ID
- OCIS_SYSTEM_USER_API_KEY

Is there a reasoning why those secrets are needed for the backup check tool? From what I understand, the tool is concepted to be run against a file directory and blobstore without any running oCIS instance, so we probably don't need all those configuration options to be checked, right?

**Comment by wkloucek** [comment]
@kobergj how hard would it be to have a flag to skip blob related checks? We could use this for validating a NFS snapshot only (without connecting to the S3 at all)

**Review by 2403905 — APPROVED** [review]
Looks good for me

**Comment by kobergj** [comment]
Merging this. Will deliver additional features in subsequent PRs. Just ping me.

**Pull request merged**